src/renderers/shared/ReactDebugTool.js

```diff
@@ -111,16 +111,18 @@ function beginLifeCycleTimer(debugID, timerType) {
   if (currentFlushNesting === 0) {
     return;
   }
-  warning(
-    !currentTimerType || lifeCycleTimerHasWarned,
-    'There is an internal error in the React performance measurement code. ' +
-    'Did not expect %s timer to start while %s timer is still in ' +
-    'progress for %s instance.',
-    timerType,
-    currentTimerType || 'no',
-    (debugID === currentTimerDebugID) ? 'the same' : 'another'
-  );
-  lifeCycleTimerHasWarned = true;
+  if (!lifeCycleTimerHasWarned) {
+    warning(
+      !currentTimerType,
+      'There is an internal error in the React performance measurement code. ' +
+      'Did not expect %s timer to start while %s timer is still in ' +
+      'progress for %s instance.',
+      timerType,
+      currentTimerType || 'no',
+      (debugID === currentTimerDebugID) ? 'the same' : 'another'
+    );
+    lifeCycleTimerHasWarned = true;
```

Comment:
No, this is still incorrect.
Let me clarify: `warning()` doesn’t actually produce a warning unless **its first argument is `false`**.

This code is not a warning path by itself.

```js
var isEverythingBad = false;
warning(!isEverythingBad, 'Everything is bad');
```

will not produce a warning. I know it’s a bit confusing but if you think of `warning()` as of something like `assert()`, it will make sense. The first argument is the condition you expect to be `true`.

In the current code, you set `lifeCycleTimerHasWarned` *in any case*, regardless of whether the warning was actually printed. So it gets silenced on the *first check*, not on the *first warning*.

This should work instead:

```js
if (currentTimerType) {
  warning(
    lifeCycleTimerHasWarned,
    ...
  );
  lifeCycleTimerHasWarned = true;
}
```

Alternatively:

```js
if (currentTimerType && !lifeCycleTimerHasWarned) {
  warning(
    false,
    ...
  );
  lifeCycleTimerHasWarned = true;
}
```